exec/cpg.c

```diff
@@ -734,11 +741,13 @@ static int notify_lib_joinlist(
 	qb_list_for_each(iter, &cpg_pd_list_head) {
 			struct cpg_pd *cpd = qb_list_entry (iter, struct cpg_pd, list);
 			if (mar_name_compare (&cpd->group_name, group_name) == 0) {
-				assert (joined_list_entries <= 1);
 				if (joined_list_entries) {
-					if (joined_list[0].pid == cpd->pid &&
-						joined_list[0].nodeid == api->totem_nodeid_get()) {
-						cpd->cpd_state = CPD_STATE_JOIN_COMPLETED;
+					int i;
+					for (i = 0; i < joined_list_entries; i++) {
+						if (joined_list[i].pid == cpd->pid &&
+						    joined_list[i].nodeid == api->totem_nodeid_get()) {
+							cpd->cpd_state = CPD_STATE_JOIN_COMPLETED;
+						}
 					}
```

Comment:
I am not 100% sure whether this breaks some state transition logic in edge cases - would be very glad for a verification by someone with more CPG knowledge than me ;)

Reply:
I think next line `if (cpd->cpd_state == CPD_STATE_JOIN_COMPLETED ||` may be problem (if list is not sorted) because let's say there are 3 cpds (pid 1,2 and 3) and join list has a pid 3, 1, then callback will be send more times than needed and/or not sent to some cpd:

Reply:
@Fabian-Gruenbichler sorry, ignore the previous comment.

Reply:
as per our discussion on IRC, I reworked this part to now iterate in three phases:
1. for the state transition (only if joinlist is passed)
2. for the notification/ipc dispatch
3. for left list handling (only if leftlist is passed)